addons/google_gmail/tests/test_google_gmail.py

```diff
@@ -0,0 +1,75 @@
+from odoo.tests.common import TransactionCase
+from unittest import mock
+from datetime import datetime
+from freezegun import freeze_time
+
+
+class TestIrMailServer(TransactionCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.mail_server = cls.env["ir.mail_server"].create(
+            {
+                "name": "Gmail",
+                "smtp_host": "fake.host",
+                "google_gmail_access_token": "fake_access_token",
+            }
+        )
+
+    def test_generate_oauth2_string_token(self):
+
+        cases = [
+            (
+                "2021-12-15 10:59:50",
+                False,
+                "fake_access_token",
+                ("Google Gmail: reuse existing access token. Expire in %i minutes", 0),
+            ),
+            (
+                "2021-12-15 10:59:55",
+                True,
+                "new-access-token",
+                ("Google Gmail: fetch new access token. Expires in %i minutes", 60),
+            ),
+            (
+                "2021-12-15 11:00:01",
+                True,
+                "new-access-token",
+                ("Google Gmail: fetch new access token. Expires in %i minutes", 60),
+            ),
+        ]
+
+        for (
+            current_datetime,
+            assert_new_token_generation_called,
+            expected_token,
+            expected_log,
+        ) in cases:
+            with self.subTest(currenct_datetime=current_datetime), freeze_time(
+                current_datetime
+            ), mock.patch(
+                "odoo.addons.google_gmail.models.google_gmail_mixin._logger.info"
+            ) as mock_logger, mock.patch(
+                "odoo.addons.google_gmail.models.google_gmail_mixin.GoogleGmailMixin._fetch_gmail_access_token",
+                return_value=(
+                    "new-access-token",
+                    int(datetime(2021, 12, 15, 12, 0, 1).timestamp()),
+                ),
+            ) as mock_fetch_gmail_access_token:
```

Comment:
Using overarching parentheses and inlining what can be seems more readable to me
```suggestion
with (
    self.subTest(currenct_datetime=current_datetime),
    freeze_time(current_datetime),
    mock.patch("odoo.addons.google_gmail.models.google_gmail_mixin._logger.info") as mock_logger,
    mock.patch(
        "odoo.addons.google_gmail.models.google_gmail_mixin.GoogleGmailMixin._fetch_gmail_access_token",
        return_value=("new-access-token", int(datetime(2021, 12, 15, 12, 0, 1).timestamp()))
    ) as mock_fetch_gmail_access_token
):
```

Reply:
done